Attempt to reconnect to DTD automatically if the connection drops

DanTup opened the pull request:
@elliette interested in your thoughts here.

The bug at https://github.com/Dart-Code/Dart-Code/issues/5808 is that after sleeping the machine, the DTD connection drops and the sidebar stops working. There's no handling of errors or connection drops, so the sidebar continues to look fine, but any click results in an unhandled error written to the console.

This change handles the `done` event in the DTD connection manager and automatically tries to reconnect (up to 5 times, with exponential backoff) as long as `disconnect`/`dispose` wasn't called explicitly. Without the retries, we might try before the network connection is back up (`ERR_NETWORK_IO_SUSPENDED`) - I was going to look to see if the network presence APIs are available in Dart, but actually they're not what we need - we don't need to know if you have an internet connection, because we're just connecting to localhost. So a retry seemed like the best option.

With this change, I can resume from sleep and the sidebar automatically reconnects (and re-initializes its state).

<img width="1165" height="843" alt="Screenshot 2025-12-15 184952" src="https://github.com/user-attachments/assets/b956b52d-0d21-47f8-a55a-67bd5961eee8" />

This also seems to fire for the property editor and I'm unsure whether it could replace your previous fix at https://github.com/flutter/devtools/issues/9028. You mentioned in that chat that `done` didn't fire - but I don't know if you just meant it was delayed until the machine woke up or not (in my testing, it _is_ firing, because my code is running).

Unfortunately I don't know of a way we could easily automate a test for this, but if you have ideas, let me know.

Fixes https://github.com/Dart-Code/Dart-Code/issues/5808
Fixes https://github.com/flutter/devtools/issues/9531

jakemac53 commented:
Yeah if the whole UI already doesn't appear when in this state then I think this change makes sense, but it would be nice to get a reconnect button in the disconnected UI eventually.

DanTup commented:
Ok, I redid a lot of this so it should be reviewed again as a whole.

I extracted some smaller unrelated parts from this PR into https://github.com/flutter/devtools/pull/9589 to try and reduce the size of this, but the DTDManager changes are not trivial.

I made the reconnecting overlay more general and use it as an overlay for any type of connection. It will now show progress like "Connecting...", "Reconnecting in Xs", and "Connection Failed". When the connection has failed all retries, there is a "Retry" button for the user to restart another set of retries.

Changes include:

- Repurposes the property editor "ReconnectingOverlay" as a more general "Connecting Overlay" also for the sidebar.
- If a DTD connection drops, it will be retried up to 5 times (and the new connection sent to the ValueNotifier)
- If the connection fails after all retries, a "Retry" button is shown to allow the user to restart connection attempts again
- The periodic check for the connection being alive is now part of DTDManager instead of only the property editor
- DTDManager now also emits state values so consumers can handle each state differently (rather than just connected/not connected)
- Mock editor for Stager apps now has a "Drop Connection" button to similate DTD connection drops
- Stager apps that use the mock editor now use a real DTDManager (since they connect to real DTD) and can simulate slow and failed connections
- Removed the force refresh for reconnect, and instead just made the `_DtdConnectedScreen` use a `key` so that when the connection changes, the widget gets new state

@elliette once you're happy with the code, I think it would be useful for you to re-test the connection issue you'd triggered with the property editor because I don't think we've seen exactly the same behaviour (I couldn't reproduce `done` not firing, for ex). I have tested the PE in both VS Code and the stager app and believe it's working the same as it was before (but the handling is more generic).

elliette commented on packages/devtools_app_shared/lib/src/service/dtd_manager.dart:
```diff
@@ -18,39 +23,205 @@ class DTDManager {
+        );
+        while (delay > 0) {
+          _connectionState.value = WaitingToRetryDTDState(delay);
+          await Future.delayed(const Duration(seconds: 1));
```
should this be `seconds: delay`?

elliette commented on packages/devtools_app_shared/lib/src/service/dtd_manager.dart:
```diff
@@ -18,39 +23,205 @@ class DTDManager {
+  ///
+  /// Disables automatic reconnect unless [disableReconnect] is `false`.
+  @visibleForTesting
+  Future<void> disconnectImpl({bool disableReconnect = true}) async {
```
Could we positively word this (i.e. `{bool enableReconnect = false}`)? (To avoid the double negative on line 166 above)

DanTup replied:
Done! I went with `allowReconnect` and made it clear on the docs that this doesn't _enable_ a reconnect, it only _allows_ it (eg., if there was a dispose of explicit disconnect call that triggered the call here, it still wouldn't reconnect).

elliette commented on packages/devtools_app_shared/lib/src/service/dtd_manager.dart:
```diff
@@ -18,39 +23,205 @@ class DTDManager {
+            disconnectBeforeConnecting: false,
+          );
+        }
+      }));
```
Could we pull everything in `.then` into a named method?

Also consider using `safeUnawaited` (here and elsewhere): https://github.com/flutter/devtools/blob/36af80ea9379f1f7a445e11264a48889bba3eaa9/packages/devtools_app/lib/src/shared/utils/utils.dart#L361-L369

DanTup replied:
Extracted to a method `_reconnectAfterDroppedConnection`.

> Also consider using `safeUnawaited` (here and elsewhere):

It looks like this is defined in `devtools_app` so I don't think it can be used here. Can/should we move it to `devtools_app_shared`?

DanTup replied:
Do you mean just an empty handler? If so, done. If not, what should I do?

I could only find this one `unawaited` added in this PR, but lmk if you think I missed any.

elliette replied:
Empty handler is fine, thanks!

elliette commented on packages/devtools_app_shared/lib/src/service/dtd_manager_connection_state.dart:
```diff
@@ -0,0 +1,33 @@
+// Copyright 2023 The Flutter Authors
```
Copyright 2025

elliette commented:
> @elliette some odd failures on the bots like this:
> 
> ```
> FlutterDriveProcess - Unhandled exception:
> FlutterDriveProcess - TimeoutException (500): timeout: Timed out receiving message from renderer: 299.773
> FlutterDriveProcess -   (Session info: chrome=143.0.7499.40)
> FlutterDriveProcess - #0      parseW3cResponse (package:webdriver/src/handler/w3c/utils.dart:87:9)
> FlutterDriveProcess - #1      W3cNavigationHandler.parseNavigateToResponse (package:webdriver/src/handler/w3c/navigation.dart:26:5)
> FlutterDriveProcess - #2      AsyncRequestClient.send (package:webdriver/src/common/request_client.dart:96:32)
> ```
> 
> Have you seen anything like this before? I don't _think_ they're related to my change, but I'll do some digging if they haven't been seen elsewhere.

Hmm I don't recall seeing that before so it's possible it is related to this PR and how we connect to DTD in our integration tests. We can see if they continue to fail on CI or if it was just a flake. The integration tests should all have comments saying how to run them, e.g.:

https://github.com/flutter/devtools/blob/628b84260835513ccffdd640a3fc8e1134907dd5/packages/devtools_app/integration_test/test/live_connection/app_test.dart#L12
